File: algo/ValidateBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class ValidateBinarySearchTree {
    // Approach 1: Recursive Traversal with Valid Range
    // Time Complexity: O(N)
    // Space Complexity: O(N)
    boolean min = true;
    boolean max = true;

    public boolean isValidBST(TreeNode root) {
        return dfs(root, null, null);
    }

    public boolean dfs(TreeNode root, Integer start, Integer end) {
        if (root == null) {
            return true;
        }
        if ((start != null && start >= root.val) ||
                (end != null && end <= root.val)) {
            return false;
        }
        return dfs(root.left, start, root.val) && dfs(root.right, root.val, end);
    }

}

File: algo/ValidateBinarySearchTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return dfs(root, NULL, NULL);
    }
    
private: 
    bool dfs(TreeNode* root, TreeNode* start, TreeNode* end) {
        if (root == NULL) {
            return true;
        }
        if ((start != NULL && root->val <= start->val) ||
            (end != NULL && root->val >= end->val)) {
            return false;
        }
        return dfs(root->left, start, root)
            && dfs(root->right, root, end);
    }
};